### PR TITLE
Add module for managing L7 firewall rules on MR devices

### PR DESCRIPTION
**Pull request by joshuajcoronado**
This module allows users to manage l7 firewall rules on an MR device using the /networks/{networkId}/wireless/ssids/{number}/firewall/l7FirewallRules API. 

I am unable to run integration tests, so I can't verify that I got every base covered, i used the existing mx l7 and mr l3 firewall rules modules as a base

### CODE REVIEW

**Comment by kbreit** [comment]
Did you copy the MX module and rename the proper sections to MR?

**Comment by y0rune** [comment]
I can not test in the Meraki Sandbox :( I do not have permission to create a network :(
```
yorune@MacBook-Pro ~/git/ansible_collections/cisco/meraki joshuajcoronado/meraki_mr_l7_firewall $ ansible-test network-integration --allow-unsupported meraki_mr_l7_firewall
Running meraki_mr_l7_firewall integration test role
[WARNING]: running playbook inside collection cisco.meraki

PLAY [meraki] ******************************************************************

TASK [meraki_mr_l7_firewall : Create network] **********************************
[WARNING]: Platform darwin on host localhost is using the discovered Python
interpreter at /opt/homebrew/opt/python@3.11/bin/python3.11, but future
installation of another Python interpreter could change the meaning of that
path. See https://docs.ansible.com/ansible-
core/2.14/reference_appendices/interpreter_discovery.html for more information.
fatal: [localhost]: FAILED! => {"ansible_facts": {"discovered_interpreter_python": "/opt/homebrew/opt/python@3.11/bin/python3.11"}, "changed": false, "msg": 403, "response": "OK (unknown bytes)", "status": 403, "url": "https://api.meraki.com/api/v1/organizations/549236/networks"}

TASK [meraki_mr_l7_firewall : Delete wireless SSID] ****************************
fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'net' is undefined. 'net' is undefined\n\nThe error appears to be in '/Users/yorune/git/ansible_collections/cisco/meraki/tests/output/.tmp/integration/meraki_mr_l7_firewall-312m3ipq-ÅÑŚÌβŁÈ/tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml': line 465, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  always:\n    - name: Delete wireless SSID\n      ^ here\n"}

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=2    skipped=0    rescued=0    ignored=0

NOTICE: To resume at this test target, use the option: --start-at meraki_mr_l7_firewall
FATAL: Command "ansible-playbook meraki_mr_l7_firewall-u4yihm7y.yml -i inventory.networking" returned exit status 2.
```

File: tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml [hunk 1]
```diff
@@ -0,0 +1,480 @@
+# Test code for the Meraki modules
+
+# Copyright: (c) 2022, Joshua Coronado (@joshuajcoronado) <joshua@coronado.io>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+---
+- name: Block of tasks
+  block:
+  - name: Create network
```
Comment (kbreit) [inline tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml:8]: This task needs to register a variable since it's used in the next task.

**Review by kbreit — CHANGES_REQUESTED** [review]

**Comment by joshuajcoronado** [comment]
@kbreit changes pushed

**Comment by y0rune** [comment]
Today, I have tested the playbook but again I do not have permission on the claiming new devices. So I wrote the email to dev support.

If I may ask how did you test @joshuajcoronado the module?

**Comment by joshuajcoronado** [comment]
@y0rune i tested the module locally, using my own networks. I was unable to get the integration tests working, and those tests are heavily borrowed from the mr l3 and mx l7 tests.

**Comment by y0rune** [comment]
@joshuajcoronado amazing! I do not have meraki for my own. When I know the way to test it in the dev env I am going to write here. Thank your for answer!

**Review by kbreit — CHANGES_REQUESTED** [review]

File: tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+      net_name: TestNet
+      state: present
+      type: wireless
+    register: new_net
+
+  - name: Set fact new_net
+    ansible.builtin.set_fact:
+      net: '{{ new_net.data.id }}'
+
+  - name: Query firewall rules
```
[the remark this answers is not in the document]
Comment (joshuajcoronado) [reply tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml:29]: Added

File: tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+        - policy: deny
+          type: port
+          port: 8080
+    register: multiple_rules
+
+  - name: Debug multiple_rules
+    ansible.builtin.debug:
+      var: multiple_rules
+
+  - name: Assert multiple_rules.data.rules | length == 3
```
Comment (kbreit) [inline tests/integration/targets/meraki_mr_l7_firewall/tasks/tests.yml:416]: This is failing as the length isn't matching. The output I see is below as well.

```
        "data": {
            "rules": [
                {
                    "application_category": {
                        "id": "meraki:layer7/category/27",
                        "name": "Advertising"
                    },
                    "policy": "deny",
                    "type": "applicationCategory"
                },
                {
                    "policy": "deny",
                    "port": "8080",
                    "type": "port"
                }
            ]
        },
```

```
TASK [meraki_mr_l7_firewall : Assert multiple_rules.data.rules | length == 3] ***
fatal: [localhost]: FAILED! => {
    "assertion": "multiple_rules.data.rules | length == 3",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

**Review by kbreit — CHANGES_REQUESTED** [review]

**Comment by joshuajcoronado** [comment]
okay, was able to get the integration tests runnin locally, passing now

**Comment by joshuajcoronado** [comment]
@kbreit this is ready too

**Comment by kbreit** [comment]
> @kbreit this is ready too

Just released 2.12.0 as you posted this. I can do a 2.13.0 easily enough of course.

**Comment by joshuajcoronado** [comment]
this and https://github.com/CiscoDevNet/ansible-meraki/pull/384 would be great

**Comment by kbreit** [comment]
This passed integration test and looks good so I am merging. Again, thank you so much for you contribution.

**Pull request merged**